quickwit-indexing/src/source/push_api_source.rs

```diff
@@ -0,0 +1,355 @@
+    pub previous_offset: u64,
+    pub current_offset: u64,
+    pub num_docs_processed: u64,
+}
+
+pub struct PushApiSource {
+    params: PushApiSourceParams,
+    push_api_mailbox: Mailbox<PushApiService>,
+    counters: PushApiSourceCounters,
+}
+
+impl PushApiSource {
+    pub async fn make(
+        params: PushApiSourceParams,
+        push_api_mailbox: Mailbox<PushApiService>,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let partition_id = PartitionId::from(params.index_id.clone());
+        let offset = if let Some(Position::Offset(offset_str)) =
+            checkpoint.position_for_partition(&partition_id).cloned()
+        {
+            offset_str.parse::<u64>()?
+        } else {
+            0
+        };
+
+        ensure_queue_exist_for_index(&push_api_mailbox, params.index_id.clone()).await?;
+        let push_api_source = PushApiSource {
+            params,
+            push_api_mailbox,
+            counters: PushApiSourceCounters {
+                previous_offset: offset,
+                current_offset: offset,
+                num_docs_processed: 0,
+            },
+        };
+        Ok(push_api_source)
+    }
+}
+
+/// Creates a queue for the index if it does not exists.
+async fn ensure_queue_exist_for_index(
+    push_api_mailbox: &Mailbox<PushApiService>,
+    index_id: String,
+) -> anyhow::Result<()> {
+    let queue_exists_req = QueueExistsRequest {
+        queue_id: index_id.clone(),
+    };
+    if push_api_mailbox
+        .ask_for_res(queue_exists_req)
+        .await
+        .map_err(|err| anyhow::anyhow!(err.to_string()))?
+    {
+        return Ok(());
+    }
+
+    let create_queue_req = CreateQueueRequest { queue_id: index_id };
+    push_api_mailbox
+        .ask_for_res(create_queue_req)
+        .await
+        .map_err(|err| anyhow::anyhow!(err.to_string()))
+}
+
+#[async_trait]
+impl Source for PushApiSource {
+    async fn emit_batches(
+        &mut self,
+        batch_sink: &Mailbox<Indexer>,
+        ctx: &SourceContext,
+    ) -> Result<Option<Duration>, ActorExitStatus> {
+        let start_after = if self.counters.current_offset == 0 {
+            None
+        } else {
+            Some(self.counters.current_offset)
+        };
+
+        let fetch_req = FetchRequest {
+            index_id: self.params.index_id.clone(),
+            start_after,
+            num_bytes_limit: self
+                .params
+                .batch_num_bytes_threshold
+                .or(Some(BATCH_NUM_BYTES_THRESHOLD)),
+        };
+        let fetch_resp = self
+            .push_api_mailbox
+            .ask_for_res(fetch_req)
+            .await
+            .map_err(|err| anyhow::anyhow!(err.to_string()))?;
+
+        let FetchResponse {
+            first_position: first_position_opt,
+            doc_batch: doc_batch_opt,
+        } = fetch_resp;
+
+        if first_position_opt.is_none() {
+            // TODO: We are currently waiting for a constant time.
```

Comment:
can we comment about what this if-statement.

Is it what happens when the fetch response is empty and there no message available?